### DISCUSSION TITLE
Mock navigation in Storybook 7 with Next.js 13 App Router

### DISCUSSION THREAD

**Discussion by StephenSHorton** [storybookjs/storybook#25470] [Help] [answered]
## How do I simulate navigation and capture the new pathname in a component?
I have been made aware of the zero-config nextjs/storybook integration. It appears the next/navigation is 'mocked' and I can manipulate it in my story like so
```ts
parameters: {
  nextjs: {
    appDirectory: true,
    navigation: {
      pathname: '/foo',
    },
  },
},
```

I have tried to use `usePathname` from `next/navigation`
```ts
const pathname = usePathname()
const currentPage = pathname.split('/').pop()
  ```
  
 But I've found out that due to navigation being mocked I can't access changes to the pathname through `usePathname`. Is there any way I can latch onto the changes in pathname and affect the state of my component to present in my Story?

### Additional information

Using Storybook version `^7.6.6`

Here's my `.storybook > main.ts`
```ts
import type { StorybookConfig } from '@storybook/nextjs'
import path from 'path'

const config: StorybookConfig = {
  stories: ['../src/**/*.mdx', '../src/**/*.stories.@(js|jsx|mjs|ts|tsx)'],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-onboarding',
    '@storybook/addon-interactions',
    '@storybook/addon-themes',
    '@storybook/themes',
  ],
  framework: {
    name: '@storybook/nextjs',
    options: { builder: { useSWC: true } },
  },
  docs: {
    autodocs: 'tag',
  },
  typescript: {
    reactDocgen: 'react-docgen',
  },
  webpackFinal: async (config) => {
    if (!config.resolve) return config
    config.resolve.alias = {
      ...config.resolve.alias,
      '@': path.resolve(__dirname, '../src'),
    }
    return config
  },
}
export default config

```

**Comment by 2wheeh** [comment]
Hi, I'm facing similar situation (https://github.com/storybookjs/storybook/discussions/26084). Have you found out any solution for this?

**Reply by StephenSHorton** [reply]
No, sorry..

**Comment by qcatch** [comment]
Im using the app router and initially had the same Issue, but I managed to get it working. Heres what I did

```
export const Default: Story = {
  parameters: {
    nextjs: {
      appDirectory: true,
      navigation: {
        pathname: '/games'
      }
    }
  },
  args: { ... }
};
```


and in my navigation components (I'm using a `use client`) and 
```
const pathname = usePathname();
const currentPath = pathname?.slice(1);
```

now `currentPath` appears as `games` in storybook

**Reply by StephenSHorton** [reply]
Thanks for replying! For context though, I was looking to simulate navigation. Meaning I'd like to click a next.js link and see the path change to another value. (For example, to show off active link styles on a navbar)

If yours does this please let me know.

**Reply by johnschult** [reply]
@qcatch This was perfect! I am actually using a server component in a story that uses `next-intl` hooks and this fixed the issue 😄

**Comment by kagankan** [comment, marked as answer]
I was also struggling with the same issue, but it was resolved with Storybook 8.1.


```tsx:.storybook/preview.tsx
import type { Preview } from "@storybook/react";
import {
  ReadonlyURLSearchParams,
  getRouter,
  usePathname,
  useSearchParams,
} from "@storybook/nextjs/navigation.mock";
import { useMemo } from "react";
import mockRouter from "next-router-mock";

const preview: Preview = {
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
  beforeEach: () => {
    getRouter().push.mockImplementation(
      (...args: Parameters<typeof mockRouter.push>) => mockRouter.push(...args)
    );
    getRouter().replace.mockImplementation(
      (...args: Parameters<typeof mockRouter.replace>) =>
        mockRouter.replace(...args)
    );
    usePathname.mockImplementation(() => mockRouter.pathname);
    useSearchParams.mockImplementation(() => {
      const searchParams = useMemo(
        () =>
          new ReadonlyURLSearchParams(
            new URLSearchParams(mockRouter.query as Record<string, string>)
          ),
        [mockRouter.query]
      );
      return searchParams;
    });
  },
};

export default preview;
```

**Reply by girlwebdeveloper** [reply]
thanks. this one works upon initial installation of storybook

**Reply by gauravvermatechno** [reply]
great this works but i have to refresh then i can see path change effects

**Reply by AxelOord** [reply]
@gauravvermatechno did you by any chance find a solution for that?
